[AOTInductor] Enforce no_grad for Run entries

SherlockNoMad opened the pull request:
Summary:
Always enter no_grad mode in AOTInductor run entries.

```
// AOTInductor uses at::addmm_out, which doesn't supports
// arguments that requires gradient. For this reason, we
// enforce no_grad context for run APIs.
```

Test Plan:
buck2 test mode/dev-nosan caffe2/test/inductor:test_aot_inductor

and OSS CI

Differential Revision: D50432042




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

khabinov commented on torch/_inductor/codegen/aoti_runtime/interface.cpp:
```diff
@@ -26,6 +26,17 @@
+    bool prev_mode = aoti_torch_grad_mode_is_enabled(); \
+    aoti_torch_grad_mode_set_enabled(false);            \
+    __VA_ARGS__                                         \
+    aoti_torch_grad_mode_set_enabled(prev_mode);        \
```
Wondering if we need some sort of RAII to restore the previous grad mode in case an exception occurs?

chenyang78 reviewed (approved):
I think Oleg made a good point regarding recovering from Exceptions. LGTM otherwise. Thanks!

desertfire commented on torch/_inductor/codegen/aoti_runtime/interface.cpp:
```diff
@@ -26,6 +26,17 @@
+// enforce no_grad context for run APIs.
+#define WITH_NO_GRAD(...)                               \
+  do {                                                  \
+    bool prev_mode = aoti_torch_grad_mode_is_enabled(); \
```
Consider print a warning msg if `prev_mode` is true.

SherlockNoMad replied:
This will be invoked in critical path of run(), so I shouldn't print log at every run() call. 

I can print warning at ContainerCreate(), but there is a chance that create() is called under no_grad, but run() is not.

SherlockNoMad commented:
@pytorchbot merge

SherlockNoMad commented:
@pytorchbot merge

SherlockNoMad committed:
[AOTInductor] Enforce no_grad for Run entries (#111613)

Summary:

Always enter no_grad mode in AOTInductor run entries.

```
// AOTInductor uses at::addmm_out, which doesn't supports
// arguments that requires gradient. For this reason, we
// enforce no_grad context for run APIs.
```

Test Plan:
buck2 test mode/dev-nosan caffe2/test/inductor:test_aot_inductor

and OSS CI

Reviewed By: khabinov, chenyang78

Differential Revision: D50432042

SherlockNoMad commented:
@pytorchbot merge
